src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -131,14 +142,34 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
 
         public async ValueTask DisposeAsync()
         {
-            switch (Services)
+            // The user didn't change the ContentRootFileProvider instance, we can dispose it
+            if (ReferenceEquals(_hostEnvironment.ContentRootFileProvider, _defaultProvider))
+            {
+                // Dispose the content provider
+                await DisposeAsync(_hostEnvironment.ContentRootFileProvider).ConfigureAwait(false);
+            }
+            else
             {
-                case IAsyncDisposable asyncDisposable:
-                    await asyncDisposable.DisposeAsync().ConfigureAwait(false);
-                    break;
-                case IDisposable disposable:
-                    disposable.Dispose();
-                    break;
+                // In the rare case that the user replaced the ContentRootFileProvider, dispose it and the one
```

Comment:
For my education, how can this happen? Setting it on the `HostBuilderContext.HostingEnvironment.ContentRootFileProvider` during a `ConfigureAppConfiguration` callback?

Reply:
It's settable 😄